### PR TITLE
lib raster: fixed security vulnerabilities and weaknesses

### PR DESCRIPTION
**Pull request by jadenabrams100**
This PR fixes three vulnerabilities/weaknesses found with older scans of Coverity. 

Issue 1208372 in error.c concerns an unbounded read of an environment variable into memory. An attacker could overwrite the environment variable that is accessed by G__home() and exploit it to overflow the buf array.

Issue 1501330 in mapset_msc.c concerns writing into an array that is not null terminated. If the path variable was not null terminated, the write could fill the whole array with data without a null terminator, causing trouble down the line.

Issue 1207344 in raster/r.gwflow/main.c concerns a constant variable guarding dead code. This is not exactly a security vulnerability, but is a code quality issue I was able to easily fix.

### CODE REVIEW

File: lib/gis/mapset_msc.c [hunk 1]
```diff
@@ -183,8 +183,9 @@ int G_make_mapset_object_group_basedir(const char *type, const char *basedir)
 int make_mapset_element_impl(const char *p_path, const char *p_element,
                              bool race_ok)
 {
-    char path[GPATH_MAX], *p;
+    char path[GPATH_MAX + 1], *p;
     const char *element;
+    path[GPATH_MAX] = 0;
```
Comment (marisn) [inline lib/gis/mapset_msc.c:188]: Although I do not think here not-null terminated p_path can reach this function, it wouldn't harm either.

Please change from 0 to '\0' to make your intent clearer.

**Review by marisn — CHANGES_REQUESTED** [review]

**Review by marisn — DISMISSED** [review]

**Review by nilason — CHANGES_REQUESTED** [review]
I added some suggestion. In general it is advisable to constrain buffer sizes defined by constants like `GPATH_MAX` to include null terminator (i.e. not to increase it with 1).

File: lib/gis/error.c [hunk 2]
```diff
@@ -368,9 +368,9 @@ void G_init_logging(void)
 
     logfile = getenv("GIS_ERROR_LOG");
     if (!logfile) {
-        char buf[GPATH_MAX];
+        char buf[GPATH_MAX + 1];
```
[the remark this answers is not in the document]
Comment (jadenabrams100) [reply lib/gis/error.c]: Ok I just added that in. Thank you!

File: lib/gis/mapset_msc.c [hunk 3]
```diff
@@ -183,8 +183,9 @@ int G_make_mapset_object_group_basedir(const char *type, const char *basedir)
 int make_mapset_element_impl(const char *p_path, const char *p_element,
                              bool race_ok)
 {
-    char path[GPATH_MAX], *p;
+    char path[GPATH_MAX + 1], *p;
```
[the remark this answers is not in the document]
Comment (jadenabrams100) [reply lib/gis/mapset_msc.c]: Ok I have added that in. Thank you!

**Comment by jadenabrams100** [comment]
I am not certain why the MacOS job failed.

File: lib/gis/mapset_msc.c [hunk 4]
```diff
@@ -183,14 +183,16 @@
 int make_mapset_element_impl(const char *p_path, const char *p_element,
                              bool race_ok)
 {
-    char path[GPATH_MAX], *p;
+    char path[GPATH_MAX] = {'\0'};
+    char *p;
     const char *element;
+    path[GPATH_MAX] = '\0';
```
[the remark this answers is not in the document]
Comment (nilason) [reply lib/gis/mapset_msc.c:189]: You should delete line 189.

**Comment by echoix** [comment]
> I am not certain why the MacOS job failed.

Just to make sure I made a rerun of the job, and still failed early.

The earliest possible error that I was only able to spot was the linker not being able to link libgrass, but didn’t find another more descriptive error in the compilation.

**Comment by echoix** [comment]
I'm surprised that another Clang-based job didn't fail. I only see that Travis builds with gcc and clang, but on Ubuntu focal. I thought we had more clang jobs than that.

**Comment by nilason** [comment]
> I am not certain why the MacOS job failed.

In the CI log you can find the reason:
```
mapset_msc.c:189:5: error: array index 4096 is past the end of the array (that has type 'char[4096]') [-Werror,-Warray-bounds]
    path[GPATH_MAX] = '\0';
    ^    ~~~~~~~~~
mapset_msc.c:186:5: note: array 'path' declared here
    char path[GPATH_MAX] = {'\0'};
    ^
1 error generated.
```

which is the same issue as the CodeQL reports.

**Comment by nilason** [comment]
> I'm surprised that another Clang-based job didn't fail. I only see that Travis builds with gcc and clang, but on Ubuntu focal. I thought we had more clang jobs than that.

The macOS CI is built with `-Wall -Wextra -Wpedantic -Werror`, which makes the difference. The gcc CI's have the same build flags, but gcc has different set of warnings activated with those compared to clang.

File: lib/gis/mapset_msc.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (jadenabrams100) [reply lib/gis/mapset_msc.c:189]: I do not know how that slipped past my mind, I will get that fixed my apologies!

**Comment by echoix** [comment]
Is this ready now that CI passes?

**Comment by jadenabrams100** [comment]
> Is this ready now that CI passes?

Yes, it is ready now.

**Review by nilason — APPROVED** [review]

**Pull request merged**